server/tailsql/options.go

```diff
@@ -551,12 +558,22 @@ func (d *DBSpec) countFields() (n int) {
 }
 
 func (d *DBSpec) checkValid() error {
-	switch {
-	case d.Source == "":
-		return errors.New("missing source")
-	case d.Driver == "":
+	if d.Source == "" {
+		return errors.New("missing source name")
+	}
+
+	// Case 1: A programmatic data source.
+	if d.DB != nil {
+		if d.countFields() != 0 {
+			return errors.New("exactly one connection source must be set")
+		}
+		return nil
+	}
+
+	// Case 2: A database/sql database.
+	if d.Driver == "" {
 		return errors.New("missing driver name")
-	case d.countFields() != 1:
+	} else if d.countFields() != 1 {
```

Comment:
drop else?

Reply:
I think it's nicer to read with the whole check combined. It makes sense to separate them when they're unrelated, but here they're part of the same check.